pkg/sync/v16/client.go

```diff
@@ -204,14 +204,17 @@ func (s *sync) write(o *unstructured.Unstructured) error {
 
 		o.SetResourceVersion(rv)
 		_, err = dc.Resource(res, o.GetNamespace()).Update(o)
-		if err != nil && strings.Contains(err.Error(), "updates to parameters are forbidden") {
-			s.log.Infof("object %s is not updateable, will delete and re-create", o.GetName())
-			err = dc.Resource(res, o.GetNamespace()).Delete(o.GetName(), &metav1.DeleteOptions{})
-			if err != nil {
-				return
+		if err != nil {
+			if strings.Contains(err.Error(), "updates to parameters are forbidden") ||
+				(strings.Contains(err.Error(), "field is immutable") && strings.Contains(o.GetName(), "omsagent")) {
```

Comment:
while I agree this would minimize the potential impact, I see no harm in dropping the ``o.GetName() == "omsagent"`` conditional, the original conditional was a doing a blanket check based on the error message.

Reply:
I think I agree with this -- @jim-minter is there any risk in always deleting and recreating sync-managed objects that have immutable field changes?

Reply:
I am not very fussed about v3 and I am okay to remove the extra check for `omsagent`.
`It is generally discouraged to make label selector updates and it is suggested to plan your selectors up front.` -  https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates
In my opinion, going forward if we do need to change spec.selectors we need to be aware that this is a bigger change than for example adding a label or changing resources. 
I agree that broken upgrade cycle is high price to pay, but I think this should be solved by adding tests, not by making exceptions wider. 
(we have an upgrade test, why didn't it fail?)

Reply:
`strings.Contains(err.Error(), "field is immutable") && strings.Contains(o.GetName(), "omsagent")` feels a fraction tight but presumably works for now to get unblocked.

`strings.Contains(err.Error(), "field is immutable")` possibly feels a bit loose.

I think I'd ship as is.  Failing that I think I'd either try to fix on a longer error message and/or kind DaemonSet.

Reply:
Answering my own question from above - the upgrade test did not fail because they don't run with azure monitoring enabled.